chapters/en/Unit 13 - Outlook/Retention.mdx

```diff
@@ -0,0 +1,352 @@
+
+- **Manhattan Self-Attention (MaSA)**
+![Attention Comparison](https://huggingface.co/datasets/hf-vision/course-assets/resolve/main/Attention%20Comparison.png)
+MaSA incorporates Self-Attention mechanism with a two-dimensional bidirectional spatial decay matrix based on the Manhattan distance among the tokens. This matrix decreases attention scores for tokens further away from a target token, allowing it to perceive global information while varying attention based on distance.
+
+- **Decomposed Manhattan Self-Attention (MaSAD)**
+![MaSAD](https://huggingface.co/datasets/hf-vision/course-assets/resolve/main/MaSAD.png)
+This mechanism decomposes Self-Attention in images along horizontal and vertical axes of the image, maintaining the spatial decay matrix without losing prior information. This decomposition allows the Manhattan Self-Attention (MaSA) to model global information efficiently with linear complexity, while preserving the original MaSA's receptive field shape.
+
+However, unlike the original RetNet that performs training with parallel representation and inference with recurrent representation, RMT does both processes with MaSA mechanism. The authors have done comparisons between MaSA and other RetNet's representations, and they show that MaSA has the best throughput with the highest accuracy.
+![MaSA vs Retention](https://huggingface.co/datasets/hf-vision/course-assets/resolve/main/MaSA%20vs%20Retention.png)
+
+### ViR
+![ViR](https://huggingface.co/datasets/hf-vision/course-assets/resolve/main/ViR.png)
+
+Another work inspired by the RetNet architecture is the ViR, as discussed in the paper [ViR: Vision Retention Networks](http://arxiv.org/abs/2310.19731). In this architecture, the authors propose a general vision backbone with a redesigned retention mechanism. They demonstrate that ViR can scale favorably to larger image resolutions in terms of image throughput and memory consumption by leveraging the dual parallel and recurrent properties of the retentive network.
+
+The overall architecture of ViR is quite similar to that of ViT, except that it replaces the Multi-Head Attention (MHA) with Multi-Head Retention (MHR). This MHR mechanism is free of any gating function and can be switched between parallel, recurrent, or chunkwise (a hybrid between parallel and recurrent) modes. Another difference in ViR is that the positional embedding is first added to the patch embedding, and then the [class] token is appended.
+
+## PyTorch Implementation of Multi-Scale Retention
+The implementation of Multi-Scale Retention (main proposal in RetNet) from the original paper is shown below:
+```python
+import torch
+import torch.nn.functional as F
+from torch import nn
+from .rms_norm import RMSNorm
```

Comment:
I think we need to have an independent code snippet. These imports from the original code will be confusing to readers.

Reply:
I see, I'll add the rest of the code so that it becomes independent. Thanks for the suggestion!

Reply:
Sounds cool!